tom/changelog.py

```diff
@@ -0,0 +1,153 @@
+                branch_name_parts = branch.split('.')
+                version_parts = [
+                        branch_name_parts[0],
+                        branch_name_parts[1],
+                        '0'
+                        ]
+        return '.'.join(version_parts)
+
+    def get_changelog_for(self, name, arg, old_version, branch):
+        olddir = os.getcwd()
+        try:
+            os.chdir(os.path.join(self.repos_root, name))
+            cmd = ['../core/misc/changelog-generator/changelog-generator',
+                    arg,
+                    '{}...{}'.format(old_version, branch)]
+            log.debug('running command: {}'.format(' '.join(cmd)))
+            proc = subprocess.run(cmd,
+                    stdout=subprocess.PIPE,
+                    universal_newlines=True,
+                    check=True)
+        finally:
+            os.chdir(olddir)
+        return proc.stdout
+
+    def generate_changelog_in_repo(self, branch, repo):
+        name = repo.repo_name
+        changelog_filename = self.changelog_filename(name)
+        if not changelog_filename:
+            # no need to generate changelog in _this_ repo
+            return False
+        # find old version
+        (header, old_version, old_changelog) = self.split_changelog_into_parts(changelog_filename, repo)
+        new_version = self.get_next_version(old_version, branch)
+        log.debug('next version: {} => {}'.format(old_version, new_version))
+        # checkout new branch
+        timestamp = re.sub('[^0-9-]', '_', str(datetime.datetime.today()))
+        new_branchname = '{}-changelog-{}'.format(new_version, timestamp)
+        repo.checkout(new_branchname, True)
+        if header != '':
+            header += '\n'
+        if name != 'enterprise':
+            # generate changelog only for this repo
+            new_changelog = self.get_changelog_for(name, '--repo', old_version, branch)
+            log.debug('new changelog: [{}]'.format(new_changelog))
+            changelog_contents = header+new_version+':\n'+new_changelog+'\n'+old_changelog+'\n'
+        else:
+            changelog_enterprise = self.get_changelog_for(name, '--enterprise', old_version, branch)
+            log.debug('enterprise changelog:'+changelog_enterprise)
+            changelog_buildscripts = self.get_changelog_for('buildscripts', '--repo', old_version, branch)
+            log.debug('buildscripts changelog:'+changelog_buildscripts)
+            changelog_contents = (new_version+':\n'+changelog_enterprise+'\n'+
+                    '        Packaging changes:\n'+changelog_buildscripts+'\n'+old_changelog+'\n')
+        repo.put_file(changelog_filename, changelog_contents)
+        repo.commit('Added changelog for '+new_version)
+        repo.push(new_branchname)
+        pr_text = self.github.create_pr(
+            'cfengine/'+name, branch, self.username, new_branchname,
+            'Changelog for ' + branch, '')
+        return pr_text
```

Comment:
Please add spaces around the `+`s above and also some blank lines to separate some logical blocks. Otherwise this is very hard to read.